packages/babel/src/utils.js

```diff
@@ -14,3 +14,15 @@ export function warnOnce(ctx, msg) {
 
 const regExpCharactersRegExp = /[\\^$.*+?()[\]{}|]/g;
 export const escapeRegExpCharacters = (str) => str.replace(regExpCharactersRegExp, '\\$&');
+
+export function stripHashAndQuery(id) {
+  const [withoutHash, hash] = id.split('#');
```

Comment:
lately, another use edge case has popped up: https://github.com/rollup/plugins/issues/528 . I believe that @rollup/plugin-maintainers & @lukastaegert should express their opinion about both of those things, or we should have util that would handle this in the same manner in all plugins interested in resolving & similar.

Maybe in here we could search for the last #/? (keeping in mind that `?foo=1#bar` is also a possibility), strip that and test if it was put right after an extension? That would be somewhat more bullet proof.

Reply:
Case like `file?foo=1#bar` is already covered by [this test](https://github.com/rollup/plugins/blob/fd6c1e537c6135002d3b64fe5a5a5401925b5d3b/packages/babel/test/fixtures/with-query-and-hash/main.js#L5).

Reply:
Right - the question is: is the logic in node-resolve plugin correct now, given the mentioned #528?

Reply:
It's not correct if we want to handle the `#` is file or directory names. Is that a valid name on the file system? I don't think it is for URLs.

Reply:
Yeah, unfortunately `#` is a valid character for file paths, examples in the wild: https://www.npmjs.com/package/es5-ext#list-of-es6-shims

Reply:
In fact, `es5-ext` is an extremely common dependency, and Rollup fails for many other top NPM dependencies because of the hash translation issue. See test suite at: https://github.com/TeleworkInc/rollup-node-testing

Reply:
How about we don't strip hash here (a.k.a. treat `#` as normal path character)?  So that we can correctly handle import path like `'es5-ext/array/#/concat'`.

Further more, we can extract this strip path logic into rollup-plugin-utils, so that plugin-commonjs and plugin-node-resolve can share it.

 If you agree on this, I will modify this PR.
@Andarist @christiantjl @LarsDenBakker

Reply:
I think that will break the use case of using query params or hashes in module imports. What if the splitting logic checks if the characters occurs before or after a file extension?

Reply:
Splitting logic is needed in two different places:
- Resolve plugin(like rollup-plugin-node-resolve) resolve **import path** into **module id**
  - Resolve plugin want to pass along query and hash
- Other plugin want to get some information from **module id**
  - Example1, rollup-plugin-babel want to extract information about file extension from **module id**
  - Example2, [moduleId with query + rollup plugin enablue us to import thing with special effect](https://github.com/rollup/plugins/issues/500#issuecomment-669839428)
  - Example3, a [load plugin](https://rollupjs.org/guide/en/#load) can decide how/where to load the module according to query an hash in **module id**

Reply:
> I think that will break the use case of using query params or hashes in module imports. What if the splitting logic checks if the characters occurs before or after a file extension?

File extension often doesn't exist in **import path**. (.e.g `import mod from './my/utils#some-hash'`)

Reply:
Not often, but it is a valid import path in the browser. Anyhow I'm ok with special casing this, it's much less frequent than query parameters.